src/clarity-icons/clarity-icons-element.ts

```diff
@@ -55,25 +59,99 @@ const setIconSize = function(element: any, size: string) {
 };
 
 ClarityIconElement.prototype.connectedCallback = function() {
+    // One thing to note here is that the attributeChangedCallback method is called for every attribute first
+    // before this connectedCallback method called only once when the custom element is inserted into the DOM.
+    // So we could check whether the attribute values really changed or not.
+    // If not, we don't need to execute the same codes again.
     const host = this as HTMLElement;
 
     if (host.hasAttribute("shape")) {
-        generateIcon(host, host.getAttribute("shape"));
+        const shapeAttrValue = host.getAttribute("shape").split(/\s/)[0];
+
+        if (shapeAttrValue === this.currentShapeAttrVal) {
+            return;
+        }
+
+        this.currentShapeAttrVal = shapeAttrValue;
+
+        if (ClarityIconsApi.instance.has(this.currentShapeAttrVal)) {
+            this.currentShapeTemplate = ClarityIconsApi.instance.get(this.currentShapeAttrVal);
+        } else {
+            console.error(`'${this.currentShapeAttrVal}' is not found in the Clarity Icons set.`);
+            host.innerHTML = ClarityIconsApi.instance.get("error");
+            return;
+        }
+    }
+
+    if (host.hasAttribute("title")) {
+        const titleAttrValue = host.getAttribute("title");
+
+        if (this.currentTitleAttrVal !== titleAttrValue) {
+            this.currentTitleAttrVal = titleAttrValue;
+        }
+
+        if (!this.currentShapeAttrVal) {
+            return;
+        }
```

Comment:
Why is this here, in the middle of the title code? If the icon has a title but no shape, you're never setting its size, but if the title has no title and no shape you do set the size correctly?

Reply:
This is here to escape calling `this.injectTemplate(host)` as a user hasn't yet provided the shape at this point.